File: src/main/java/io/felipeandrade/gsw2/block/ColorProviderMaterialBlock.java
package io.felipeandrade.gsw2.block;

import net.fabricmc.fabric.api.client.rendering.v1.ColorProviderRegistry;
import net.minecraft.block.BlockState;
import net.minecraft.client.color.block.BlockColorProvider;
import net.minecraft.client.color.item.ItemColorProvider;
import net.minecraft.item.ItemStack;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.BlockRenderView;

import io.felipeandrade.gsw2.common.ClientResourceInit;
import io.felipeandrade.gsw2.material.GSWMaterial;

public class ColorProviderMaterialBlock extends GSWMaterialBlock
        implements BlockColorProvider, ItemColorProvider, ClientResourceInit {


    public ColorProviderMaterialBlock(GSWMaterial material, Settings settings) {
        super(material, settings);
    }

    @Override
    public int getColor(BlockState state, BlockRenderView view, BlockPos pos, int tintIndex) {
        return material.getSecondaryColor();
    }

    @Override
    public int getColor(ItemStack stack, int tintIndex) {
        return material.getSecondaryColor();
    }

    @Override
    public void onClientInit() {
        ColorProviderRegistry.BLOCK.register(this, this);
        ColorProviderRegistry.ITEM.register(this, this);
    }

}

File: gsw-mod/src/main/kotlin/io/felipeandrade/gsw/block/ColorProviderMaterialBlock.kt
package io.felipeandrade.gsw.block

import io.felipeandrade.gsw.common.ClientResourceInit
import io.felipeandrade.gsw.material.GSWMaterial
import net.fabricmc.fabric.api.client.rendering.v1.ColorProviderRegistry
import net.minecraft.block.BlockState
import net.minecraft.client.color.block.BlockColorProvider
import net.minecraft.client.color.item.ItemColorProvider
import net.minecraft.item.ItemStack
import net.minecraft.util.math.BlockPos
import net.minecraft.world.BlockRenderView

class ColorProviderMaterialBlock(
    gswMaterial: GSWMaterial,
    settings: Settings,
) : GSWMaterialBlock(gswMaterial, settings), BlockColorProvider, ItemColorProvider, ClientResourceInit {

    override fun getColor(state: BlockState, view: BlockRenderView?, pos: BlockPos?, tintIndex: Int): Int =
        gswMaterial.secondaryColor

    override fun getColor(stack: ItemStack, tintIndex: Int): Int {
        return gswMaterial.secondaryColor
    }

    override fun onClientInit() {
        ColorProviderRegistry.BLOCK.register(this, this)
        ColorProviderRegistry.ITEM.register(this, this)
    }
}